collector/sysctl_bsd.go

```diff
@@ -50,28 +55,66 @@ type bsdSysctl struct {
 	dataType bsdSysctlType
 
 	// Post-retrieval conversion hooks
-	conversion func(uint64) uint64
+	conversion func(float64) float64
 }
 
 func (b bsdSysctl) Value() (float64, error) {
 	var tmp32 uint32
 	var tmp64 uint64
+	var tmpf64 float64
 	var err error
 
 	switch b.dataType {
 	case bsdSysctlTypeUint32:
 		tmp32, err = unix.SysctlUint32(b.mib)
-		tmp64 = uint64(tmp32)
+		tmpf64 = float64(tmp32)
 	case bsdSysctlTypeUint64:
 		tmp64, err = unix.SysctlUint64(b.mib)
+		tmpf64 = float64(tmp64)
+	case bsdSysctlTypeStructTimeval:
+		raw, err := unix.SysctlRaw(b.mib)
+		if err != nil {
+			return 0, err
+		}
+
+		/*
+		 * From 10.3-RELEASE sources:
+		 *
+		 * /usr/include/sys/_timeval.h:47
+		 *  time_t      tv_sec
+		 *  suseconds_t tv_usec
+		 *
+		 * /usr/include/sys/_types.h:60
+		 *  long __suseconds_t
+		 *
+		 * ... architecture dependent, via #ifdef:
+		 *  typedef __int64_t __time_t;
+		 *  typedef __int32_t __time_t;
+		 */
+		if len(raw) != (C.sizeof_time_t + C.sizeof_suseconds_t) {
+			// Shouldn't get here, unless the ABI changes...
+			return 0, fmt.Errorf(
+				"Length of bytes recieved from sysctl (%d) does not match expected bytes (%d).",
+				len(raw),
+				C.sizeof_time_t+C.sizeof_suseconds_t,
+			)
+		}
+
+		seconds_up := unsafe.Pointer(&raw[0])
+		suseconds_up := uintptr(seconds_up) + C.sizeof_time_t
+		unix := int64(*(*C.time_t)(seconds_up))
```

Comment:
I think both are valid options. Would say let's keep it and have high precision. The surface is limited so that's fine.